orderer/common/throttle/ratelimit.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package throttle
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// Clock observes time
+type Clock interface {
+	Now() time.Time
+	After(d time.Duration) <-chan time.Time
+}
+
+type SharedRateLimiter struct {
+	// Time is how this rate limiter observes time
+	Time Clock
+	// Limit is the total limit of throughput for all clients
+	Limit int
+	// InactivityTimeout is the inactivity time after which a client is pruned from memory
+	InactivityTimeout time.Duration
+	// state
+	once    sync.Once
+	stop    chan struct{}
+	lock    sync.RWMutex
+	clients map[string]*clientState
+}
+
+type clientState struct {
+	lock       sync.Mutex
+	lastSeen   atomic.Value
+	budget     chan struct{}
+	budgetUsed uint32
+}
+
+func (srl *SharedRateLimiter) LimitRate(client string) {
+	srl.once.Do(srl.init)
+
+	cs := srl.getOrCreateState(client)
+
+	srl.lock.RLock()
+	numClients := len(srl.clients)
+	srl.lock.RUnlock()
```

Comment:
Maybe make a separate getLen method to improve readability? Above there is a call to getOrCreateState method inside which srl.lock is also handled.

Reply:
done